dao-voting/src/main.sw

```diff
@@ -0,0 +1,295 @@
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor has not been called to initialize
+    /// - The user deposits an asset that is not the specified governance token.
+    /// - The user does not deposit and assets
+    #[storage(read, write)]fn deposit() -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(storage.gov_token == msg_asset_id(), Error::NotGovernanceToken);
+        require(msg_amount() > 0, Error::NoAssetsSent);
+
+        let sender: Identity = sender_identity();
+
+        let prev_balance = storage.balances.get(sender);
+        let new_balance = prev_balance + msg_amount();
+        storage.balances.insert(sender, new_balance);
+
+        true
+    }
+
+    /// Update the user balance to indicate they have withdrawn governance tokens
+    ///
+    /// # Parameters
+    ///
+    /// amount - amount of governance tokens to withdraw from the contract
+    ///
+    /// # Panics
+    ///
+    /// This functions will panic when:
+    /// - The constructor has not been called to initalize
+    /// - The user tries to withdraw more than their balance
+    #[storage(read, write)]fn withdraw(amount: u64) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+
+        let sender: Identity = sender_identity();
+
+        let prev_balance = storage.balances.get(sender);
+        require(prev_balance >= amount, Error::NotEnoughAssets);
+
+        let new_balance = prev_balance - amount;
+        storage.balances.insert(sender, new_balance);
+
+        // Transfer the asset back to the user
+        match sender {
+            Identity::Address(address) => transfer_to_output(amount, storage.gov_token, address), Identity::ContractId(address) => force_transfer_to_contract(amount, storage.gov_token, address), 
+        }
+
+        true
+    }
+
+    /// Vote on a given proposal
+    ///
+    /// # Parameters
+    ///
+    /// proposal_id - proposal to vote on
+    /// vote_amount - amount of votes to use on proposal
+    /// is_yes_vote - determines if you vote yes or no on the proposal
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor has not been called to initialize
+    /// - The proposal id is out of range
+    /// - The vote amount is 0
+    /// - The vote amount is greater than the users deposited balance
+    /// - The given proposal is expired
+    #[storage(read, write)]fn vote(proposal_id: u64, vote_amount: u64, is_yes_vote: bool) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(proposal_id < storage.proposal_count, Error::InvalidId);
+        require(vote_amount > 0, Error::VoteAmountCannotBeZero);
+
+        let sender: Identity = sender_identity();
+        let sender_balance = storage.balances.get(sender);
+
+        require(sender_balance >= vote_amount, Error::NotEnoughAssets);
+
+        let mut proposal = storage.proposals.get(proposal_id);
```

Comment:
I would change the logic to first fetch the proposal and check the deadline and then the amount for voting